src/lib/scheduler/worker.cpp

```diff
@@ -128,17 +139,55 @@ void Worker::join() {
 uint64_t Worker::num_finished_tasks() const { return _num_finished_tasks; }
 
 void Worker::_wait_for_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) {
-  auto tasks_completed = [&tasks]() {
-    // Reversely iterate through the list of tasks, because unfinished tasks are likely at the end of the list.
-    for (auto it = tasks.rbegin(); it != tasks.rend(); ++it) {
-      if (!(*it)->is_done()) {
+  // This lambda checks if all tasks from the vector have been executed. If they are, it causes _wait_for_tasks to
+  // return. If there are remaining tasks, it primarily tries to execute these. If they cannot be executed, the
+  // worker performs work for others (i.e., executes tasks from the queue).
+  auto check_own_tasks = [&]() {
+    auto all_done = true;
+    for (auto it = tasks.begin(); it != tasks.end(); ++it) {
+      const auto& task = *it;
+      if (task->is_done()) {
+        continue;
+      }
+
+      // Task is not yet done - check if it is ready for execution
+      if (!task->is_ready()) {
+        all_done = false;
+        continue;
+      }
+
+      // Give other tasks a certain chance of being executed, too. Anectotal evidence says that this is a good idea.
+      // For some reason, this keeps the memory consumption of TPC-H Q6 low even if the scheduler is overcommitted.
+      // Because generating random numbers is somewhat expensive, we keep a list of random numbers and reuse them.
+      // TODO(anyone): Look deeper into scheduling theory and make this theoretically sound.
+      _next_random = (_next_random + 1) % _random.size();
+      if (_random[_next_random] <= 20) {
```

Comment:
How certain are you about that random number? I am wondering if we are too often exiting here for many small tasks (TPC-C).

Reply:
For posterity: @Bouncner is re-running benchmark_all with 50 right now